walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,65 @@
 
 package com.google.sps;
 
+import java.util.Comparator;
+import java.util.List;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+
 
 public final class FindMeetingQuery {
+
+ /* Given the meeting information, this method
+    returns the times when the meeting could happen that day.
+ */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> options = new ArrayList<TimeRange>();
+    if (Integer.compare((int) request.getDuration(), TimeRange.WHOLE_DAY.duration()) > 0) {
+        // If duration of event is more than 24hrs
+        return options;
+    }
+
+    if (events.equals(Collections.emptySet())) {
+        // If there are no events scheduled
+        options.add(TimeRange.WHOLE_DAY);
+        return options;
+    }
+    
+    // Most recent start time for the chunk of meeting time option
+    int currentStart = 0;
+
+    List<Event> eventsList = new ArrayList<Event>(events);
+    // eventsList.sort(Comparator.comparing(Event::getWhen));
+    // Iterate through event list in order
+    System.out.println("eventsList.size(): " + eventsList.size());
+    for (int i = 0; i < eventsList.size(); i++) {
+        System.out.println("i: " + i);
+        if (Collections.disjoint(eventsList.get(i).getAttendees(), request.getAttendees())) {
+            // If event attendees and request attendees have no names in common
+            continue;
+        }
+
+        if (eventsList.get(i).getWhen().start() < currentStart) {
+            // If current event overlaps with an event that's already going on and it's start time < currentStart
+            currentStart = eventsList.get(i).getWhen().end() + 1;
+            continue;
+        }
+
+        if ((eventsList.get(i).getWhen().start() - request.getDuration()) >= currentStart) {
+            // If there's time before event i starts, add it as a chunk of meeting time options
+            options.add(TimeRange.fromStartEnd(currentStart, eventsList.get(i).getWhen().start(), false));
```

Comment:
It doesn't seem like you are ever deleting from this `options`. There is a case when this would fail.

You can add a TimeRange (12-1pm) to `options` but then later in the eventsList, there may be an event with TimeRange(12:30-12:45pm). As a result, you would need to delete/check again TimeRange (12-1pm) in `options`

This seems like a hard and inefficient approach. I think a smarter approach maybe to grab all the potential conflicting TimeRanges that matter from eventsList, and then iterate over all the conflicting TimeRanges to determine the working TimeRanges